Thread safe dft

jdodinh opened the pull request:
The use of `RefCell` in many DFT implementations prevents sharing across threads (`RefCell` is `!Sync`). This has been noted in the code several times:
https://github.com/Plonky3/Plonky3/blob/9200da320029aa7b3dee281c1a0af89bf788d428/dft/src/radix_2_small_batch.rs#L42-L44
https://github.com/Plonky3/Plonky3/blob/9200da320029aa7b3dee281c1a0af89bf788d428/monty-31/src/dft/mod.rs#L49-L51

This PR makes the DFTs thread-safe by placing internal state (e.g., twiddle caches) behind `Arc<RwLock<…>>`. It also adds a simple Rayon-based benchmark that runs multiple batched DFTs in parallel.

SyxtonPrime commented:
Yeah this looks like a reasonable approach. Will review once the tests pass.

jdodinh commented:
I see, this is unfortunate. I look into another solution.

SyxtonPrime reviewed (commented):
Hmm, this seems to be leading to a lot of allocation calls. (You've thrown `to_vec` in quite a few places).

As far as I can tell, most of these don't seem to be necessary. I've commented on a few of them explicitly but this comment applies to basically all of them.

Is there a reason you don't want to work with slices and similar?

jdodinh commented on monty-31/src/dft/mod.rs:
```diff
@@ -125,10 +122,29 @@ impl<MP: FieldParameters + TwoAdicData> RecursiveDft<MontyField31<MP>> {
+                if fft_len > cur_have {
+                    *inv_tw_lock = new_inv_twiddles; // move in the new table
+                }
+            }
```
Yes, exactly, the idea here is that it could've been updated in the meantime. I think that the write lock should be avoided as much as possible.

I did a significant refactor of this section in 70a5a90124133d9a61a050ab8d81f1be7981a0b0.

jdodinh commented on monty-31/src/dft/mod.rs:
```diff
@@ -125,10 +122,29 @@ impl<MP: FieldParameters + TwoAdicData> RecursiveDft<MontyField31<MP>> {
+    }
+    fn get_inv_twiddles(&self) -> Vec<Vec<MontyField31<MP>>> {
+        self.inv_twiddles.read().to_vec()
+    }
```
I changed this in 70a5a90124133d9a61a050ab8d81f1be7981a0b0.

jdodinh commented:
Pushed some updates, but still not done. Will request review when done.

jdodinh commented:
Just to clarify: the bug isn’t in `main`; it’s introduced by this PR. I changed `RecursiveDft` to make twiddles append-only (so we don’t recompute them when growing FFT sizes). That requires mirroring indices in several places (e.g., `a[a.len() - 1 - i]` instead of `a[i]`), which exposed issues the current tests didn’t catch.

I realize this mixes two concerns (append-only twiddles + thread safety). Would you prefer I split the index-mirroring changes into a separate PR, or at least add tests for larger `RecursiveDft` instances?

SyxtonPrime reviewed (commented):
I'm still confused why you need to double `Arc` and add lots of clones.

For example in radix_2_small_batch, if I make the set of changes I suggest here, we seem to be able to remove one of the Arc's and all of the clones.

Is there a reason we don't like this strategy?

Assuming this looks good, can you make the same changes to the other fft files? (radix_2_dit, radix_2_dit_parallel, dft/mod)

PS: Even if you want to keep the `Arc`'s in, I think you can still get rid of all the clones.

SyxtonPrime commented on monty-31/src/dft/mod.rs:
```diff
@@ -98,36 +95,66 @@ impl<MP: FieldParameters + TwoAdicData> RecursiveDft<MontyField31<MP>> {
-            self.twiddles.replace(new_twiddles);
-            self.inv_twiddles.replace(new_inv_twiddles);
+        // let curr_max_fft_len = 2 << self.twiddles.read().len();
+        let need = log2_strict_usize(fft_len);
+        if self.twiddles.read().len() + 1 >= need {
+            return;
+        }
+
+        let snapshot = self.twiddles.read().clone();
+        let have = snapshot.len() + 1;
+        if have >= need {
+            return;
+        }
```
These are both doing the same check.

jdodinh replied:
Deleted in 34f59e226a6c32e08a0ef96de41125523f6f2da8.

SyxtonPrime commented on monty-31/src/dft/mod.rs:
```diff
@@ -98,36 +95,66 @@ impl<MP: FieldParameters + TwoAdicData> RecursiveDft<MontyField31<MP>> {
+        let missing_inv_twiddles: Vec<Vec<MontyField31<MP>>> = missing_twiddles
+            .iter()
+            .map(|ts| {
+                let mut v = Vec::with_capacity(ts.len());
```
While this is probably marginally faster, the code is already fast enough so the simpler iterator methods using `collect` is probably better.

Also in general `F::zero_vec()` will be faster than using `vec::with_capacity`.

jdodinh replied:
Addressed in 34f59e226a6c32e08a0ef96de41125523f6f2da8

SyxtonPrime reviewed (commented):
2 other quick comments

jdodinh commented:
Thanks for the review! Let me explain the layering and why I kept the inner `Arc` plus the “clone-then-drop-lock” pattern.

Why two Arcs?
- Outer `Arc<RwLock<…>>`: lets the DFT struct be cheaply clonable (`RwLock` itself isn’t `Clone`) and shared across threads.
- Inner Arc<[…]>: gives readers a snapshot of the twiddle table that they can access without keeping a read lock for the whole FFT. Cloning this Arc is just a refcount bump—no deep copy—so it’s cheap.

The concurrency concern this solves

Consider `A` wants an FFT of size `n` (twiddles already present) while `B` needs size `2n` (requires growth).
- With a plain `Vec` inside the `RwLock`, `A` would typically read-lock and borrow `&twiddles[…]` for the entire FFT. That keeps the read lock held, so `B` can’t acquire a write lock to grow the table until `A`’s FFT finishes.
With `Arc<[…]>`, `A` does:
```
let snap = self.twiddles.read().clone(); // clone the Arc, drop the lock immediately
```
It then runs the FFT using `snap` (no lock held).
- Now `B` can compute missing layers off-lock, then take a brief write lock to replace the inner `Arc` with the grown table. Existing workers keep using the old `Arc`; when they finish, the old table is dropped automatically (refcount goes to 0).

So: the “clone” here is intentional and cheap—its sole purpose is to release the read lock early while keeping stable, immutable access to the data.

I hope this makes sense, and please feel free to correct me if you think there's anything wrong with my rationale.

jdodinh commented on dft/src/radix_2_small_batch.rs:
```diff
@@ -39,13 +40,12 @@ pub struct Radix2DFTSmallBatch<F> {
     /// bit reversed order. The final set of twiddles `twiddles[-1]` is the
     /// one element vectors `[1]` and more general `twiddles[-i]` has length `2^i`.
     ///
```
Done in 34f59e226a6c32e08a0ef96de41125523f6f2da8.

nikkolasg commented:
Hey -  Great work by @jdodinh  ! Just signalling that we would love this PR to get in, it'd make it super useful for us where we run in a fully parallelizable context. So a gentle ping @SyxtonPrime and if there is anything we can do to help to push this to finish line 🙏

tcoratger commented on monty-31/src/dft/forward.rs:
```diff
@@ -32,6 +32,25 @@ impl<MP: FieldParameters + TwoAdicData> MontyField31<MP> {
+                v
+            })
             .collect()
     }
```
Probably this avoids the large intermediate vector by calculating the specific generator for each required FFT size and then collecting only its powers.

```suggestion
    pub fn get_missing_twiddles(req_lg_n: usize, cur_lg_n: usize) -> Vec<Vec<Self>> {
      // Get the main generator for the largest required FFT size.
      let main_generator = Self::two_adic_generator(req_lg_n);
  
      (cur_lg_n..req_lg_n)
          .map(|level| {
              // For a given 'level', we're generating twiddles for a DIF pass
              // where the number of butterflies is m = 2^level.
              let count = 1 << level;
  
              // The generator for this smaller FFT size is a power of the main generator.
              //
              // The exponent is 2^(req_lg_n - (level + 1)).
              let sub_generator_exp = 1 << (req_lg_n - level - 1);
              let sub_generator = main_generator.exp_u64(sub_generator_exp as u64);
  
              // Now, we can collect the 'count' powers of this specific sub-generator.
              sub_generator.powers().collect_n(count)
          })
          .collect()
  }
```

SyxtonPrime commented:
I think I follow you. I guess I was thinking that those clones might be somewhat expensive as I thought we were cloning the whole twiddle vec instead of just the Arc.

I'll have another look through keeping that in mind. Based on previous looks, I think we are pretty close to mergeing :)

tcoratger reviewed (approved):
LGTM

SyxtonPrime commented on baby-bear/src/baby_bear.rs:
```diff
@@ -231,13 +231,33 @@ mod tests {
+        crate::BabyBear,
+        super::EF,
+        p3_monty_31::dft::RecursiveDft<_>
+    );
```
Seems not to be a huge problem so I'm merging.

If this becomes a problem we can do back and remove/reduce these.
